### PR TITLE
[if-else] use bit operation for categorical features

### PR DESCRIPTION
**Pull request by wxchan**

### CODE REVIEW

**Comment by wxchan** [comment]
@guolinke is my current approach right? should I use one bitset for each tree or each node?

**Comment by guolinke** [comment]
@wxchan 
I think both are okay. 
For each node, the problem is need to use different variate name.

**Comment by wxchan** [comment]
@guolinke I use `Common::FindInBitset(cat_threshold_.data() + cat_boundaries_[cat_idx], cat_boundaries_[cat_idx + 1] - cat_boundaries_[cat_idx], int_fval)` instead of bitset now. You can check if it's ok.

File: src/io/tree.cpp [hunk 1]
```diff
@@ -310,6 +310,14 @@ std::string Tree::ToIfElse(int index, bool is_predict_leaf_index) {
   if (num_leaves_ == 1) {
     str_buf << "return 0";
   } else {
+    str_buf << "std::vector<uint32_t> cat_threshold = {";
```
Comment (guolinke) [inline src/io/tree.cpp]: const  ?

Comment (wxchan) [inline src/io/tree.cpp]: @guolinke should I use `constexpr`? and should I put it outside of function `PredictTree`?

Comment (guolinke) [inline src/io/tree.cpp]: it seems const variate is enough: https://stackoverflow.com/questions/4748083/when-should-you-use-constexpr-capability-in-c11

**Pull request merged**